modules/plugin/shapefile/src/main/java/org/geotools/data/shapefile/ShapefileFeatureSource.java

```diff
@@ -336,13 +327,10 @@ protected int getCountInternal(Query query) throws IOException {
             if (records != null) {
                 goodRecs = new CloseableIteratorWrapper<Data>(records.iterator());
             }
-        } else if (getDataStore().isIndexed() && !bbox.isNull() 
-                && !Double.isInfinite(bbox.getWidth()) && !Double.isInfinite(bbox.getHeight())) {
+        } else if (getDataStore().isIndexed() && filter != null) {
             try {
-                if(indexManager.isSpatialIndexAvailable() || getDataStore().isIndexCreationEnabled()) {
-                    goodRecs = indexManager.querySpatialIndex(bbox);
-                }
-            } catch (TreeException e) {
+                goodRecs = indexManager.queryFilterIndex(filter, q.getMaxFeatures());
```

Comment:
okay so perhaps this thing is handling both a bbox filter and an attribute filter, I will keep an eye out for it.